src/rust/engine/engine_pyo3/src/externs/interface/fs.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::path::PathBuf;
+
+use fs::{GlobExpansionConjunction, PathGlobs, PreparedPathGlobs, StrictGlobMatching};
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+
+pub(crate) fn register(m: &PyModule) -> PyResult<()> {
+  m.add_function(wrap_pyfunction!(match_path_globs, m)?)?;
+  Ok(())
+}
+
+struct PyPathGlobs(PathGlobs);
+
+impl PyPathGlobs {
+  fn parse(self) -> PyResult<PreparedPathGlobs> {
+    self.0.clone().parse().map_err(|e| {
+      PyValueError::new_err(format!(
+        "Failed to parse PathGlobs: {:?}\n\nError: {}",
+        self.0, e
+      ))
+    })
+  }
+}
+
+impl<'source> FromPyObject<'source> for PyPathGlobs {
+  fn extract(obj: &'source PyAny) -> PyResult<Self> {
+    let globs: Vec<String> = obj.getattr("globs")?.extract()?;
+
+    let description_of_origin_field: String = obj.getattr("description_of_origin")?.extract()?;
+    let description_of_origin = if description_of_origin_field.is_empty() {
+      None
+    } else {
+      Some(description_of_origin_field)
+    };
+
+    let match_behavior_str: &str = obj
+      .getattr("glob_match_error_behavior")?
+      .getattr("value")?
+      .extract()?;
+    let match_behavior = StrictGlobMatching::create(match_behavior_str, description_of_origin)
```

Comment:
I was kind of hoping for an API closer to `serde::Deserialize` here, where a derive macro would to most of the field-mapping work for you, and you'd be able to write something like:

```rust
let path_globs: PathGlobs = pyo3_serde::from_object(obj)?;
```

rather than having to get each field by field-name and put them together yourself... That would probably necessitate`description_of_origin` being a field on `glob_match_error_behavior` rather than a sibling field in `obj`.

I guess this is https://github.com/PyO3/pyo3/issues/566 and I definitely wouldn't suggest implementing it on the critical path of this conversion, but it would be _nice_ to avoid that boilerplate... :)

Reply:
It's also worth calling out that this wrapping of a Python object is an alternative to making the object native as we do in https://github.com/pantsbuild/pants/blob/8824cc3b157e38f72cdb89004f4955eb776dfb58/src/rust/engine/src/externs/fs.rs#L67

I'm kindof curious what making the object native would look like with `Py03`'s API... it might be less awkward than with `rust-cpython`'s, and I expect that it would be more performant than double-constructing objects.

Reply:
Agreed. I didn't go with a native class because we already have examples of that and to land the port successfully I want to avoid rewriting everything to do that. But after the port is finished, it makes sense.

Reply:
We don't actually have any examples of native objects with Py03... but maybe the macro/API is very similar?